The ``100-matrix_mul`` module
=============================

using ``matrix_mul``
--------------------

importing function
>>> lazy_matrix_mul = __import__('101-lazy_matrix_mul').lazy_matrix_mul

testing with correct input
==========================

>>> print(lazy_matrix_mul([[1, 2], [3, 4]], [[1, 2], [3, 4]]))
[[ 7 10]
 [15 22]]

>>> print(lazy_matrix_mul([[1, 2]], [[3, 4], [5, 6]]))
[[13 16]]

testing with wrong input
========================

>>> print(lazy_matrix_mul("hello", [[2, 3], [4, 5]]))
Traceback (most recent call last):
...
ValueError: Scalar operands are not allowed, use '*' instead

>>> print(lazy_matrix_mul([[1, 2]], [[6, 7], "now"]))
Traceback (most recent call last):
...
ValueError: setting an array element with a sequence

>>> print(lazy_matrix_mul([], [[2, 3], [4, 5]]))
Traceback (most recent call last):
...
ValueError: shapes (0,) and (2,2) not aligned: 0 (dim 0) != 2 (dim 0)

>>> print(lazy_matrix_mul([[1, 2]], [[6, 7], [4, 'a']]))
Traceback (most recent call last):
...
TypeError: invalid data type for einsum

>>> print(lazy_matrix_mul([[1, 2], [3, 4, -7.8]], [[1, 2], [3, 4]]))
Traceback (most recent call last):
...
ValueError: setting an array element with a sequence.

>>> print(lazy_matrix_mul([[1, 2]], [[3, 4]]))
Traceback (most recent call last):
...
ValueError: shapes (1,2) and (1,2) not aligned: 2 (dim 1) != 1 (dim 0)

>>> print(lazy_matrix_mul([[1, 2]], "hello"))
Traceback (most recent call last):
...
ValueError: Scalar operands are not allowed, use '*' instead

>>> print(lazy_matrix_mul([1, 2], [[1, 2], [3, 4]]))
[ 7 10]

>>> print(lazy_matrix_mul([[1, 2]], []))
Traceback (most recent call last):
...
ValueError: shapes (1,2) and (0,) not aligned: 2 (dim 1) != 0 (dim 0)

>>> print(lazy_matrix_mul([['a', 'b']], [[1, 2], [3, 4]]))
Traceback (most recent call last):
...
TypeError: invalid data type for einsum

>>> print(lazy_matrix_mul([[1, 2]], [[1, 2], [1, 2, 3]]))
Traceback (most recent call last):
...
ValueError: setting an array element with a sequence.

>>> print(lazy_matrix_mul([[1, 3]]))
Traceback (most recent call last):
...
TypeError: lazy_matrix_mul() missing 1 required positional argument: 'm_b'

>>> print(lazy_matrix_mul())
Traceback (most recent call last):
...
TypeError: lazy_matrix_mul() missing 2 required positional arguments: 'm_a' and 'm_b'
